packages/jellyfish-api-core/__tests__/category/blockchain/getMempoolAncestors.test.ts

```diff
@@ -0,0 +1,100 @@
+      .rejects
+      .toThrowError('RpcApiError: \'Transaction not in mempool\', code: -5, method: getmempoolancestors')
+  })
+
+  it('should return error if transaction id has an invalid length', async () => {
+    await expect(testing.rpc.blockchain.getMempoolAncestors('invalidtxidstring'))
+      .rejects
+      .toThrow('RpcApiError: \'parameter 1 must be of length 64 (not 17, for \'invalidtxidstring\')\', code: -8, method: getmempoolancestors')
+  })
+})
+
+describe('transactions with ancestors', () => {
+  const container = new MasterNodeRegTestContainer()
+  const testing = Testing.create(container)
+
+  beforeAll(async () => {
+    await testing.container.start()
+    await testing.container.waitForWalletCoinbaseMaturity()
+    for (let i = 0; i < 3; i++) {
+      await testing.rpc.wallet.sendToAddress('mwsZw8nF7pKxWH8eoKL9tPxTpaFkz7QeLU', 0.003)
+    }
+  })
+
+  afterAll(async () => {
+    await testing.container.stop()
+  })
+
+  it('should return JSON object if verbose is true', async () => {
+    const txId = await testing.rpc.wallet.sendToAddress('mwsZw8nF7pKxWH8eoKL9tPxTpaFkz7QeLU', 0.003)
+    const mempoolAncestors = await testing.rpc.blockchain.getMempoolAncestors(txId, true)
+    const keys = Object.keys(mempoolAncestors)
+    expect(keys.length).toBeGreaterThan(0)
+    for (const key of keys) {
+      expect(mempoolAncestors[key]).toStrictEqual({
+        fees: expect.any(Object),
+        vsize: expect.any(BigNumber),
+        weight: expect.any(BigNumber),
+        fee: expect.any(BigNumber),
+        modifiedfee: expect.any(BigNumber),
+        time: expect.any(BigNumber),
+        height: expect.any(BigNumber),
+        descendantcount: expect.any(BigNumber),
+        descendantsize: expect.any(BigNumber),
+        descendantfees: expect.any(BigNumber),
+        ancestorcount: expect.any(BigNumber),
+        ancestorsize: expect.any(BigNumber),
+        ancestorfees: expect.any(BigNumber),
+        wtxid: expect.any(String),
+        depends: expect.any(Array),
+        spentby: expect.any(Array),
+        'bip125-replaceable': expect.any(Boolean)
+      })
+    }
+  })
+
+  it('should return array of transaction ids if verbose is false', async () => {
+    const txId = await testing.rpc.wallet.sendToAddress('mwsZw8nF7pKxWH8eoKL9tPxTpaFkz7QeLU', 0.003)
+    const mempoolAncestors = await testing.rpc.blockchain.getMempoolAncestors(txId, false)
+    expect(mempoolAncestors.length).toBeGreaterThan(0)
+    for (const ancestorId of mempoolAncestors) {
+      expect(ancestorId).toStrictEqual(expect.stringMatching(/^[0-9a-f]{64}$/))
+    }
+  })
+
+  it('should return array of transaction ids if verbose is undefined', async () => {
+    const txId = await testing.rpc.wallet.sendToAddress('mwsZw8nF7pKxWH8eoKL9tPxTpaFkz7QeLU', 0.003)
+    const mempoolAncestors = await testing.rpc.blockchain.getMempoolAncestors(txId)
+    expect(mempoolAncestors.length).toBeGreaterThan(0)
```

Comment:
IMO testing greater than 0 here is pretty scary since we're not expecting any behaviour. Ie. Since we sent 3 transactions to chain, we should eventually see 3 ancestors here. As it is currently, even if there is just 1 ancestor, the test will pass.

We can clear the mempool by generating a block after each test so that the tests are always run with a clean slate.

Reply:
Correct me if my understanding is wrong, but as the method is getting the ancestors of a particular transaction ID, the number of ancestors of that particular transaction might not be the same as the number of transactions in the mempool
(referencing the diagram [here](https://bitcoin.stackexchange.com/questions/99168/clarifying-descendants-ancestors-in-error-cases-for-too-long-mempool-chain))
In that case, it would be difficult to assert exactly how many ancestors a transaction has (except that it is greater than 0). Also ran the tests a number of times with logs and I could not get an ancestor length of 3

Reply:
MasternodeRegtestContainer gives us a local private defichain network to test against
hence the number of transactions, when a block is mined, etc. are all deterministic

Reply:
This is my understanding:

Say Alice initiates 3 transactions to Bob and Bob initiates 1 transaction to Cat,
the tx hierarchy would look like:

```mermaid
graph LR
  A(Alice) --"0.5 (1)"--> B(Bob)
  A --"0.5 (2)"--> B
  A --"0.5 (3)"--> B
  B --"1.5 (4)"--> C(Cat)
```

Prior to the block being mined, `(4)` would have ancestors `(1) (2) (3)`

Reply:
based on my observation of the mempool, the hierachy looks something more like this

```
         Tb
        /
Ta  --- Td
        \        
          Tc 
                 \--- Te
                  \___ Tf
```
so even though there are 6 transactions in the mempool, Td will have only 1 ancestor (Ta) but Te will have 2 ancestors (Ta and Tc).
alternatively, i could assert something like: 
```
const rawMempool = await testing.rpc.blockchain.getRawMempool(true)
expect(keys.length).toStrictEqual(rawMempool[txId].ancestorcount.toNumber() - 1)
```
ref line 432@: [](https://github.com/DeFiCh/ain/blob/master/src/rpc/blockchain.cpp)

but that would assume that the getRawMempool method returns the correct number of ancestors. let me know your thoughts?

Reply:
Yup this tree looks ok too. So I would test precisely that:
- T<sub>d</sub> has 1 ancestor T<sub>a</sub>
- T<sub>e</sub> has 2 ancestors T<sub>a</sub> and T<sub>c</sub> (though I'm not sure if T<sub>c</sub> will be returned - hence the test demonstrating so would be useful 🤓)

Reply:
to address your query, yes, if T<sub>e</sub> has 2 ancestors, both T<sub>a</sub> and T<sub>c</sub> will be returned. the issue is, the above diagram is just for illustration purpose. in reality, T<sub>d</sub> could have 2 ancestors and T<sub>e</sub> could have 1 ancestor, even though T<sub>e</sub> occurred after T<sub>d</sub> (the mempool has its own sorting mechanism as transactions get added). hence it would be difficult to deterministically assert exactly how many ancestors a particular transaction (eg.  T<sub>e</sub>) has in the test, without actually referencing the current state of the mempool (eg. via methods like `testing.rpc.blockchain.getRawMempool(true)`)

Reply:
Agree that there are many possible and non-trivial ways to test this 'accurately' (There are a lot of different scenarios as well).  Looking at the [test](https://github.com/DeFiCh/ain/blob/master/test/functional/mempool_packages.py#L43) in the `ain` repo, IMO it's sufficiently covers the `getmempoolancestors` case, and we can just do a 'smoke test' in jellyfish.